Source/DafnyCore/Backends/Dafny/DafnyCodeGenerator.cs

```diff
@@ -31,16 +36,34 @@ class DafnyCodeGenerator : SinglePassCodeGenerator {
       return res;
     }
 
-    public DafnyCodeGenerator(DafnyOptions options, ErrorReporter reporter) : base(options, reporter) {
+    public DafnyCodeGenerator(DafnyOptions options, ErrorReporter reporter, bool preventShadowing = true) : base(options, reporter) {
       options.SystemModuleTranslationMode = CommonOptionBag.SystemModuleMode.Include;
       if (Options?.CoverageLegendFile != null) {
         Imports.Add("DafnyProfiling");
       }
+
+      testing = Environment.GetEnvironmentVariable("TEST_DAFNY") == "true";
+      this.preventShadowing = preventShadowing;
+    }
+
+    protected override string GetCompileNameNotProtected(IVariable v) {
+      return preventShadowing ? v.CompileName : v.SanitizedNameShadowable;
     }
 
-    public static void throwGeneralUnsupported() {
```

Comment:
Just FYI I have a branch that adds a token here, since there are several callers that have one handy, and that helps narrow down where the error came from. Is that unnecessary after this change?

Relatedly, we should change the name of the method if it no longer throws all the time. That will also steer us towards being able to keep going after more unsupported features. Perhaps just `AddUnsupported`?

Reply:
Adding a token would still be great and you can definitely do it in again.

I changed so that code emitting recoverable issues now emit two newly defined exceptions, `RecoverableInvalidOperationException` and `RecoverableUnsupportedFeatureException` so that they show a better error message indicating that the error can be recovered by using the new option.

I also rename the methods to `AddUnsupported` and `AddUnsupportedFeature`, respectively